app/grandchallenge/core/middleware.py

```diff
@@ -1,33 +1,57 @@
-from allauth_2fa.middleware import BaseRequire2FAMiddleware
-from django.urls import Resolver404, get_resolver
+from allauth import app_settings
+from allauth.account.adapter import get_adapter
+from allauth.account.urls import urlpatterns as account_urlpatterns
+from allauth.mfa.utils import is_mfa_enabled
+from allauth.socialaccount.urls import (
+    urlpatterns as social_account_urlpatterns,
+)
+from django.contrib import messages
+from django.http import HttpResponseRedirect
 from django.utils.deprecation import MiddlewareMixin
 
+from grandchallenge.subdomains.utils import reverse
 
-class RequireStaffAndSuperuser2FAMiddleware(BaseRequire2FAMiddleware):
-    def require_2fa(self, request):
-        # Staff users and superusers are required to have 2FA.
+
+class RequireStaffAndSuperuser2FAMiddleware(MiddlewareMixin):
+    """Force multi-factor authentication for staff users and superusers."""
+
+    allowed_urls = [

```

Comment:
Should we instead allow any url in the allauth app? Might be easier to maintain. Any risks to that?

Reply:
I think it would be ok to do that. It would mean that all authentication related views are always accessible, which I guess is fair. The MFA urls I didn't add are no blockers, they only work after MFA set-up, which is why I didn't add them, but there is no harm in adding them.

Reply:
I was curious about how to do this and GPT4 told me, might be handy and more robust to the frequent breaking changes in allauth:

```python
from django.urls import URLResolver, URLPattern
from django.conf import settings
from django.utils.module_loading import import_string

def list_url_names(urlconf_name, prefix=''):
    """
    Recursively lists all URLs in a URL configuration and their pattern names.
    """
    # Import the URL configuration module dynamically
    try:
        urls = import_string(urlconf_name)
    except ImportError:
        # This will handle the case where the URL configuration is not a module but an object
        urls = urlconf_name

    # Iterate through all URL patterns
    for pattern in urls.urlpatterns:
        if isinstance(pattern, URLPattern):
            # This is a final URL pattern
            pattern_name = pattern.name
            if pattern_name:
                print(f"{prefix}{pattern.pattern.regex.pattern} -> {pattern_name}")
        elif isinstance(pattern, URLResolver):
            # This pattern is an include, so recurse into the included URLconf
            new_prefix = pattern.pattern.regex.pattern
            if prefix:
                new_prefix = f"{prefix}{new_prefix}"
            list_url_names(pattern.urlconf_name, new_prefix)

# Specify the URL configuration of 'allauth.urls'
list_url_names('allauth.urls')
```

Reply:
Maybe we would also want to add a test so that we're sure the set of allowed urls is sensible, we at least then get alerted to upstream changes.

Reply:
Ah nice, I would have just done `*[pattern.name for pattern in all_auth_urls]`, importing `from allauth.account.urls import urlpatterns as account_urlpatterns` but what GPT4 proposes looks more robust and adding a test is a good idea as well.